### ISSUE TITLE
Fails to build on som 32-bit architectures: unresolved import `std::sync::atomic::AtomicU64`

### ISSUE THREAD

**Issue by jonassmedegaard** [oxigraph/oxigraph#684]
Building on some architectures, including armv5te, armv7 and i686, fails like this:
```
     Running `CARGO=/usr/bin/cargo CARGO_CRATE_NAME=oxigraph CARGO_MANIFEST_DIR=/<<PKGBUILDDIR>>/lib CARGO_PKG_AUTHORS='Tpt <thomas@pellissier-tanon.fr>' CARGO_PKG_DESCRIPTION='a SPARQL database and RDF toolkit
' CARGO_PKG_HOMEPAGE='https://oxigraph.org/' CARGO_PKG_LICENSE='MIT OR Apache-2.0' CARGO_PKG_LICENSE_FILE='' CARGO_PKG_NAME=oxigraph CARGO_PKG_REPOSITORY='https://github.com/oxigraph/oxigraph/tree/main/lib' CARGO_PKG_RUST_VERSION=1.60 CARGO_PKG_VERSION=0.3.21 CARGO_PKG_VERSION_MAJOR=0 CARGO_PKG_VERSION_MINOR=3 CARGO_PKG_VERSION_PATCH=21 CARGO_PKG_VERSION_PRE='' CARGO_PRIMARY_PACKAGE=1 LD_LIBRARY_PATH='/<<PKGBUILDDIR>>/target/release/deps:/usr/lib' rustc --crate-name oxigraph --edition=2021 lib/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -C linker-plugin-lto -C codegen-units=1 --cfg 'feature="default"' --cfg 'feature="http_client"' --cfg 'feature="oxhttp"' --cfg 'feature="rocksdb-pkg-config"' -C metadata=1508ef35c907c0b3 -C extra-filename=-1508ef35c907c0b3 --out-dir /<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/deps --target armv5te-unknown-linux-gnueabi -L dependency=/<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/deps -L dependency=/<<PKGBUILDDIR>>/target/release/deps --extern digest=/<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/deps/libdigest-3b13d90b19b04646.rmeta --extern hex=/<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/deps/libhex-fcd234fe14a16a25.rmeta --extern json_event_parser=/<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/deps/libjson_event_parser-bb3885a47b9392b3.rmeta --extern lazy_static=/<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/deps/liblazy_static-f747e84c73c57323.rmeta --extern libc=/<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/deps/liblibc-fd20330d9aad3f1f.rmeta --extern md5=/<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/deps/libmd5-4e55fe8048e8673b.rmeta --extern oxhttp=/<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/deps/liboxhttp-faed4b25f08c352e.rmeta --extern oxilangtag=/<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/deps/liboxilangtag-c338aed92553b296.rmeta --extern oxiri=/<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/deps/liboxiri-272081c5153846d4.rmeta --extern oxrdf=/<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/deps/liboxrdf-99fe898828ecfaab.rmeta --extern oxrocksdb_sys=/<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/deps/liboxrocksdb_sys-7ff3b29ed8beda55.rmeta --extern oxsdatatypes=/<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/deps/liboxsdatatypes-d1f40aac0f16ba6c.rmeta --extern rand=/<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/deps/librand-bc211c6fee2ea7b6.rmeta --extern regex=/<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/deps/libregex-ccb12da91eb8cf1d.rmeta --extern rio_api=/<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/deps/librio_api-5e8dbc566bb6a1a0.rmeta --extern rio_turtle=/<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/deps/librio_turtle-51206faf7954d581.rmeta --extern rio_xml=/<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/deps/librio_xml-abbc8e1bfc3ad386.rmeta --extern sha1=/<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/deps/libsha1-7af6f454bf22f921.rmeta --extern sha2=/<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/deps/libsha2-daabba51ab4e5b8d.rmeta --extern siphasher=/<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/deps/libsiphasher-dcea87e88c59aa8f.rmeta --extern sparesults=/<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/deps/libsparesults-6a3b47d7013008e6.rmeta --extern spargebra=/<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/deps/libspargebra-88d78c63db43ee04.rmeta -Wtrivial-casts -Wtrivial-numeric-casts -Wunsafe-code -Wunused-lifetimes -Wunused-qualifications '-Wclippy::borrow-as-ptr' '-Wclippy::case-sensitive-file-extension-comparisons' '-Wclippy::cast-lossless' '-Wclippy::cast-possible-truncation' '-Wclippy::cast-possible-wrap' '-Wclippy::cast-precision-loss' '-Wclippy::cast-ptr-alignment' '-Wclippy::cast-sign-loss' '-Wclippy::checked-conversions' '-Wclippy::clone-on-ref-ptr' '-Wclippy::cloned-instead-of-copied' '-Wclippy::copy-iterator' '-Wclippy::dbg-macro' '-Wclippy::decimal-literal-representation' '-Wclippy::default-trait-access' '-Wclippy::default-union-representation' '-Wclippy::empty-enum' '-Wclippy::enum-glob-use' '-Wclippy::exit' '-Wclippy::expect-used' '-Wclippy::expl-impl-clone-on-copy' '-Wclippy::explicit-deref-methods' '-Wclippy::explicit-into-iter-loop' '-Wclippy::explicit-iter-loop' '-Wclippy::filter-map-next' '-Wclippy::flat-map-option' '-Wclippy::fn-to-numeric-cast-any' '-Wclippy::from-iter-instead-of-collect' '-Wclippy::get-unwrap' '-Wclippy::if-not-else' '-Wclippy::if-then-some-else-none' '-Wclippy::implicit-clone' '-Wclippy::inconsistent-struct-constructor' '-Wclippy::index-refutable-slice' '-Wclippy::inefficient-to-string' '-Wclippy::inline-always' '-Wclippy::inline-asm-x86-att-syntax' '-Wclippy::inline-asm-x86-intel-syntax' '-Wclippy::invalid-upcast-comparisons' '-Wclippy::items-after-statements' '-Wclippy::large-digit-groups' '-Wclippy::large-stack-arrays' '-Wclippy::large-types-passed-by-value' '-Wclippy::let-underscore-must-use' '-Wclippy::let-unit-value' '-Wclippy::linkedlist' '-Wclippy::lossy-float-literal' '-Wclippy::macro-use-imports' '-Wclippy::manual-assert' '-Wclippy::manual-ok-or' '-Wclippy::many-single-char-names' '-Wclippy::map-unwrap-or' '-Wclippy::match-bool' '-Wclippy::match-same-arms' '-Wclippy::match-wildcard-for-single-variants' '-Wclippy::maybe-infinite-iter' '-Wclippy::mem-forget' '-Wclippy::multiple-inherent-impl' '-Wclippy::mut-mut' '-Wclippy::mutex-atomic' '-Wclippy::naive-bytecount' '-Wclippy::needless-bitwise-bool' '-Wclippy::needless-continue' '-Wclippy::needless-pass-by-value' '-Wclippy::no-effect-underscore-binding' '-Wclippy::non-ascii-literal' '-Wclippy::print-stderr' '-Wclippy::print-stdout' '-Wclippy::ptr-as-ptr' '-Wclippy::range-minus-one' '-Wclippy::range-plus-one' '-Wclippy::rc-buffer' '-Wclippy::rc-mutex' '-Wclippy::redundant-closure-for-method-calls' '-Wclippy::redundant-else' '-Wclippy::redundant-feature-names' '-Wclippy::ref-binding-to-reference' '-Wclippy::ref-option-ref' '-Wclippy::rest-pat-in-fully-bound-structs' '-Wclippy::return-self-not-must-use' '-Wclippy::same-functions-in-if-condition' '-Wclippy::single-match-else' '-Wclippy::stable-sort-primitive' '-Wclippy::str-to-string' '-Wclippy::string-add' '-Wclippy::string-add-assign' '-Wclippy::string-lit-as-bytes' '-Wclippy::string-to-string' '-Wclippy::todo' '-Wclippy::transmute-ptr-to-ptr' '-Wclippy::trivially-copy-pass-by-ref' '-Wclippy::try-err' '-Wclippy::unicode-not-nfc' '-Wclippy::unimplemented' '-Wclippy::unnecessary-self-imports' '-Wclippy::unnecessary-wraps' '-Wclippy::unneeded-field-pattern' '-Wclippy::unnested-or-patterns' '-Wclippy::unreadable-literal' '-Wclippy::unseparated-literal-suffix' '-Wclippy::unused-async' '-Wclippy::unused-self' '-Wclippy::use-debug' '-Wclippy::used-underscore-binding' '-Wclippy::verbose-bit-mask' '-Wclippy::verbose-file-reads' '-Wclippy::wildcard-dependencies' '-Wclippy::zero-sized-map-values' -C debuginfo=2 --cap-lints warn -C linker=arm-linux-gnueabi-gcc -C link-arg=-Wl,-z,relro --remap-path-prefix oxigraph_0.3.21=/usr/share/cargo/registry/oxigraph-0.3.21 --remap-path-prefix oxrdf_0.1.7=/usr/share/cargo/registry/oxrdf-0.1.7 --remap-path-prefix oxrocksdb-sys_0.3.21=/usr/share/cargo/registry/oxrocksdb-sys-0.3.21 --remap-path-prefix oxsdatatypes_0.1.3=/usr/share/cargo/registry/oxsdatatypes-0.1.3 --remap-path-prefix sparesults_0.1.8=/usr/share/cargo/registry/sparesults-0.1.8 --remap-path-prefix spargebra_0.2.8=/usr/share/cargo/registry/spargebra-0.2.8 --remap-path-prefix debian/cargo_registry=/usr/share/cargo/registry/ -L native=/<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/build/ring-7640b6290fd342eb/out -L native=/usr/lib/arm-linux-gnueabi -L native=/<<PKGBUILDDIR>>/target/armv5te-unknown-linux-gnueabi/release/build/oxrocksdb-sys-6b78f02e44e580b4/out`
error[E0432]: unresolved import `std::sync::atomic::AtomicU64`
  --> lib/src/storage/mod.rs:29:25
   |
29 | use std::sync::atomic::{AtomicU64, Ordering};
   |                         ^^^^^^^^^
   |                         |
   |                         no `AtomicU64` in `sync::atomic`
   |                         help: a similar name exists in the module: `AtomicU32`

error[E0308]: arguments to this function are incorrect
    --> lib/src/storage/backend/rocksdb.rs:1425:17
     |
1425 |         Ok(Some(min(rlimit.rlim_cur, rlimit.rlim_max)))
     |                 ^^^ ---------------  --------------- expected `u64`, found `u32`
     |                     |
     |                     expected `u64`, found `u32`
     |
help: the return type of this call is `u32` due to the type of the argument passed
    --> lib/src/storage/backend/rocksdb.rs:1425:17
     |
1425 |         Ok(Some(min(rlimit.rlim_cur, rlimit.rlim_max)))
     |                 ^^^^---------------^^^^^^^^^^^^^^^^^^
     |                     |
     |                     this argument influences the return type of `min`
help: the return type of this call is `u32` due to the type of the argument passed
    --> lib/src/storage/backend/rocksdb.rs:1425:17
     |
1425 |         Ok(Some(min(rlimit.rlim_cur, rlimit.rlim_max)))
     |                 ^^^^^^^^^^^^^^^^^^^^^---------------^
     |                                      |
     |                                      this argument influences the return type of `min`
note: function defined here
    --> /usr/src/rustc-1.70.0/library/core/src/cmp.rs:1171:14
help: you can convert a `u32` to a `u64`
     |
1425 |         Ok(Some(min(rlimit.rlim_cur.into(), rlimit.rlim_max)))
     |                                    +++++++
help: you can convert a `u32` to a `u64`
     |
1425 |         Ok(Some(min(rlimit.rlim_cur, rlimit.rlim_max.into())))
     |                                                     +++++++

Some errors have detailed explanations: E0308, E0432.
For more information about an error, try `rustc --explain E0308`.
error: could not compile `oxigraph` due to 2 previous errors
```

I guess it can be solved like this:
```patch
--- a/lib/Cargo.toml
+++ b/lib/Cargo.toml
@@ -33,6 +33,7 @@
 regex = "1"
 oxilangtag = "0.1"
 oxiri = "0.2"
+portable-atomic = "1"
 rio_api = "0.8"
 rio_turtle = "0.8"
 rio_xml = "0.8"
--- a/lib/src/storage/mod.rs
+++ b/lib/src/storage/mod.rs
@@ -26,7 +26,7 @@
 #[cfg(not(target_family = "wasm"))]
 use std::path::{Path, PathBuf};
 #[cfg(not(target_family = "wasm"))]
-use std::sync::atomic::{AtomicU64, Ordering};
+use portable_atomic::{AtomicU64, Ordering};
 #[cfg(not(target_family = "wasm"))]
 use std::sync::Arc;
 #[cfg(not(target_family = "wasm"))]
```

NB! Above patch is totally untested, but I will push it to the Debian autobuilders now, and report back it if worked as intended.

**Comment by Tpt** [comment]
Hi @jonassmedegaard 

I am not testing 32bits targets in the CI and it seems I have let slipped some errors. Sorry for that.

I am going to prepare a change to fix that.

About your last error I think you can just follow the suggestion the compiler is giving.

**Comment by Tpt** [comment]
It should be fixed in 0.4.0-alpha.2. Please reopen if it's not the case
